### PR TITLE
1d elements dev

### PR DESCRIPTION
**Pull request by termi-official**
I work coupled 1D-3D problems so for me it is sometimes helpful to visualize the solutions side by side while MFEM is computing the solution to track down issues.

The draft is not very clean yet and might come with some level of memory corruption (could not test everything in detail yet). This is basically my late try on fixing #68 .

To test the PR I currently just utilize two very simple meshes which I feed into MFEM example 0.

3D mesh
```
MFEM mesh v1.0

dimension
1

elements
5
1 1 0 1
1 1 1 2
1 1 2 3
1 1 2 4
2 1 4 5

boundary
3
1 0 0
2 0 3
3 0 5

vertices
6
3
  0    0 -1.0
1.0    0    0
2.0    0    0
3.0  1.0    0
3.0 -1.0    0
4.0 -1.0  1.0
``` 

3D vector test field
```
FiniteElementSpace
FiniteElementCollection: H1_1D_P2
VDim: 3
Ordering: 0

0
2.029082637937203
2.25675394962394
0
2.128376974811976
0
1.264541318968602
2.267918293780571
1.378376974811971
2.442565462217959
1.314188487405989
0.6947706594843015
2.179750465858888
1.880065462217956
2.412159705920951
1.783782731108984
1.709311978452903
2.293586121702257
0.7516884874059858
2.347971218514969
0.719594243702995
0
2.029082637937203
2.25675394962394
0
2.128376974811976
0
1.264541318968602
2.267918293780571
1.378376974811971
-1.442565462217959
1.314188487405989
0.6947706594843015
2.179750465858888
1.880065462217956
2.41211241205920951
1.783782731108984
1.709311978452903
-2.293586121702257
0.7516884874059858
0.347971218514969
0.719594243702995
0
2.029082637937203
2.25675394962394
0
2.128376974811976
0
1.264541318968602
2.267918293780571
1.378376974811971
2.442565462217959
-1.314188487405989
0.6947706594843015
-2.179750465858888
2.179750465858888
1.880065462217956
-2.412159705920951
1.783782731108984
2.293586121702257
-0.7516884874059858
2.347971218514969
0.719594243702995
```

2D mesh
```
MFEM mesh v1.0

dimension
1

elements
5
1 1 0 1
1 1 1 2
1 1 2 3
1 1 2 4
2 1 4 5

boundary
3
1 0 0
2 0 3
3 0 5

vertices
6
2
  0    0 
1.0    0 
2.0    0 
3.0  1.0 
3.0 -1.0 
4.0 -1.0 
```

2d vector test field
```
FiniteElementSpace
FiniteElementCollection: H1_1D_P1
VDim: 2
Ordering: 0

-0.84018773
-0.39438292
-0.78309923
-0.79844004
-0.91164738
-0.19755137
-0.33522275
-0.7682296
-0.27777472
-0.55396998
-0.47739705
-0.6288709
```

scalar test field
```
0
2.029082637937203
2.25675394962394
0
2.128376974811976
0
1.264541318968602
2.267918293780571
1.378376974811971
2.442565462217959
1.314188487405989
```

GlVis output (left 2D, right 3D)
glvis-1d-domain-example

## TODOs
* [x] Visualizing the grid without solution
* [x] Visualizing element attributes
* [ ] ~~Visualizing boundary attributes~~ (possible future work)
* [ ] ~~Visualizing the element endpoints correctly~~ (possible future work)
* [x] Visualizing high order solutions
* [x] Vector valued problems
* [ ] Merge master after #232 gets merged
* [ ] Add keys to control line width for solution interactively
* [ ] Add tests

### CODE REVIEW

**Comment by termi-official** [comment]
Turns out many of the issues I encountered has beeb fixed with MFEM patches and restructuring of the GLVis codebase. Scalar field visualization is basically working now. Some minor things remain to be resolved for scalar fields. 

However, vector-valued fields are still open, which are rather important in some areas of mechanics.

**Comment by termi-official** [comment]
Okay 3d vector visualization is also okayish now.

GLVis_s01

 The visualization is not very pretty, because modifying the renderer to include dynamic line thickness information seems a big out of scope for now. I also noticed a crazy amount of code duplication between scalar and vector field visualizer classes. Maybe a good future PR is to merge the duplicate code (and hence features).

**Comment by termi-official** [comment]
Bump @tzanio @publixsubfan . I think https://github.com/GLVis/glvis/pull/232 is not being worked on anymore? Can we merge this PR then and leave the open TODOs for a follow up PR which addresses line rendering?

**Comment by tzanio** [comment]
> Can we merge this PR then and leave the open TODOs for a follow up PR which addresses line rendering?

Yes, lets do that. Sorry for the long delay @termi-official !

**Comment by termi-official** [comment]
Absolutely no problem! :) I will merge master and fi the remaining errors tomorrow.

**Review by tzanio — APPROVED** [review]
Two more items:

1. Can you please update `CHANGELOG`

2. Should we add some of the 1D example meshes/results in MFEM and/or https://github.com/GLVis/data?

Thanks again for the contribution, @termi-official!

Tzanio

File: lib/vssolution.cpp [hunk 1]
```diff
@@ -625,15 +635,38 @@ void VisualizationSceneSolution::GetRefinedDetJ(
       }
       else if (drawelems >= 4)
       {
-         vals(j) = J.Det();
-         // if (vals(j) >= 0.0)
-         //    vals(j) = sqrt(vals(j));
-         // else
-         //    vals(j) = -sqrt(-vals(j));
```
[the remark this answers is not in the document]
Comment (termi-official) [reply lib/vssolution.cpp:632]: It is a comment on master, so it does not have any functional consequences - should I leave it anyway?

File: glvis.cpp [hunk 2]
```diff
@@ -1152,8 +1152,8 @@ int main (int argc, char *argv[])
    const char *font_name     = string_default;
    int         portnum       = 19916;
    int         multisample   = GetMultisample();
-   double      line_width    = 1.0;
-   double      ms_line_width = gl3::LINE_WIDTH_AA;
+   double      line_width    = GetLineWidth();
+   double      ms_line_width = GetLineWidthMS();
```
[the remark this answers is not in the document]
Comment (termi-official) [reply glvis.cpp:1156]: If your question is regarding the change itself, then this change just introduces a call to the internal API to keep the defaults consistent. If your question is regarding whether multisampling makes a difference, then it depends on the system. Not all systems support the respective OpenGL calls for multisampling of lines, which is part of the reason why I wanted to merge the line rendering PR before this. However, this PR is also functional without the line rendering PR.

File: lib/vssolution.cpp [hunk 3]
```diff
@@ -1354,16 +1400,14 @@ void VisualizationSceneSolution::Prepare()
          PrepareFlat2();
          return;
       default:
-         if (v_normals)
+         if (v_normals && dim > 1)
          {
             PrepareWithNormals();
```
Comment (termi-official) [inline lib/vssolution.cpp:1405]: Regarding this change here I am not super confident. This change will need intervention further intervention for the line rendering PR to setup the normals of the flat representation correctly.

**Comment by termi-official** [comment]
Sure, we can think about adding some of the 1D network mesh stuff to MFEM, but we might need to think a bit how to do this one properly. My heart simulation framework is heavy patchwork (subclasses of mesh to patch the connectivity tables) to work around the assumption that we have the scenario of more than 2 element sharing a face. I do not have a good solution to make this properly work for now, because this assumption seems to be quite deeply rooted into the MFEM core. Happy to discuss further.

File: glvis.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (termi-official) [reply glvis.cpp:1156]: For example pressing `A` on my workstation during interactive visualization removes stair-casing though antialiasing.

**Comment by tzanio** [comment]
Thanks for making the changes, @termi-official !

**Review by jandrej — APPROVED** [review]

**Pull request merged**